<h1>Data Wrangling</h1>

# TO DO
* Get the info and datatype for each column.
* Fix the typing in each column.
* Describe each dataset.
* Rename columns properly (normalize them to be lower case and with _ instead of spaces).
* Check for duplicates .
* Count the nulls in each column.
* Fill/remove corrupt data (missing values).

### Importing necessary libraries

In [ ]:
import numpy as np 
import pandas as pd
from ydata_profiling import ProfileReport

### Importing Data and reading into a Pandas DataFrame

In [28]:
cols = ['symboling', 'normalized-losses', 'make', 'fuel-type', 'aspiration', 'num-of-doors', 'body-style', 'drive-wheels', 'engine-location', 'wheel-base', 'length', 'width', 'height', 'curb-weight', 'engine-type', 'num-of-cylinders', 'engine-size', 'fuel-system', 'bore', 'stroke', 'compression-ratio', 'horsepower', 'peak-rpm', 'city-mpg', 'highway-mpg', 'price']
df = pd.read_csv('../data/Old Car Price data/imports-85.data.txt', names=cols)
print(df.shape)
df.sample(10)


(205, 26)


symboling normalized-losses           make fuel-type aspiration  \
120          1               154       plymouth       gas        std   
48           0                 ?         jaguar       gas        std   
72           3               142  mercedes-benz       gas        std   
107          0               161         peugot       gas        std   
5            2                 ?           audi       gas        std   
6            1               158           audi       gas        std   
137          2               104           saab       gas      turbo   
160          0                91         toyota       gas        std   
117          0               161         peugot       gas      turbo   
81           3               153     mitsubishi       gas        std   

    num-of-doors   body-style drive-wheels engine-location  wheel-base  ...  \
120         four    hatchback          fwd           front        93.7  ...   
48          four        sedan          rwd           front       113.0  ...   
72           two  convertible          rwd           front        96.6  ...   
107         four        sedan          rwd           front       107.9  ...   
5            two        sedan          fwd           front        99.8  ...   
6           four        sedan          fwd           front       105.8  ...   
137         four        sedan          fwd           front        99.1  ...   
160         four        sedan          fwd           front        95.7  ...   
117         four        sedan          rwd           front       108.0  ...   
81           two    hatchback          fwd           front        96.3  ...   

     engine-size  fuel-system  bore  stroke compression-ratio horsepower  \
120           90         2bbl  2.97    3.23               9.4         68   
48           258         mpfi  3.63    4.17               8.1        176   
72           234         mpfi  3.46    3.10               8.3        155   
107          120         mpfi  3.46    3.19               8.4         97   
5            136         mpfi  3.19    3.40               8.5        110   
6            136         mpfi  3.19    3.40               8.5        110   
137          121         mpfi  3.54    3.07               9.0        160   
160           98         2bbl  3.19    3.03               9.0         70   
117          134         mpfi  3.61    3.21               7.0        142   
81           122         2bbl  3.35    3.46               8.5         88   

     peak-rpm city-mpg highway-mpg  price  
120      5500       31          38   6229  
48       4750       15          19  35550  
72       4750       16          18  35056  
107      5000       19          24  11900  
5        5500       19          25  15250  
6        5500       19          25  17710  
137      5500       19          26  18620  
160      4800       38          47   7738  
117      5600       18          24  18150  
81       5000       25          32   8499  

[10 rows x 26 columns]

In [34]:
df.head()

symboling normalized-losses         make fuel-type aspiration num-of-doors  \
0          3                 ?  alfa-romero       gas        std          two   
1          3                 ?  alfa-romero       gas        std          two   
2          1                 ?  alfa-romero       gas        std          two   
3          2               164         audi       gas        std         four   
4          2               164         audi       gas        std         four   

    body-style drive-wheels engine-location  wheel-base  ...  engine-size  \
0  convertible          rwd           front        88.6  ...          130   
1  convertible          rwd           front        88.6  ...          130   
2    hatchback          rwd           front        94.5  ...          152   
3        sedan          fwd           front        99.8  ...          109   
4        sedan          4wd           front        99.4  ...          136   

   fuel-system  bore  stroke compression-ratio horsepower  peak-rpm city-mpg  \
0         mpfi  3.47    2.68               9.0        111      5000       21   
1         mpfi  3.47    2.68               9.0        111      5000       21   
2         mpfi  2.68    3.47               9.0        154      5000       19   
3         mpfi  3.19    3.40              10.0        102      5500       24   
4         mpfi  3.19    3.40               8.0        115      5500       18   

  highway-mpg  price  
0          27  13495  
1          27  16500  
2          26  16500  
3          30  13950  
4          22  17450  

[5 rows x 26 columns]

In [40]:
for column in df.columns:
    print(column, df[column].eq("?").sum())

symboling 0
normalized-losses 0
make 0
fuel-type 0
aspiration 0
num-of-doors 0
body-style 0
drive-wheels 0
engine-location 0
wheel-base 0
length 0
width 0
height 0
curb-weight 0
engine-type 0
num-of-cylinders 0
engine-size 0
fuel-system 0
bore 0
stroke 0
compression-ratio 0
horsepower 0
peak-rpm 0
city-mpg 0
highway-mpg 0
price 0


In [38]:
report = ProfileReport(df=df, title="car_profile", explorative=True)
report.to_file("./profiling reports/car_profile.html")

Summarize dataset:   0%|          | 0/5 [00:00<?, ?it/s]

/home/khaled/.cache/pypoetry/virtualenvs/introduction_to_data_analytics-jZTzfAo0-py3.10/lib/python3.10/site-packages/ydata_profiling/model/correlations.py:66: UserWarning: There was an attempt to calculate the auto correlation, but this failed.
To hide this warning, disable the calculation
(using `df.profile_report(correlations={"auto": {"calculate": False}})`
If this is problematic for your use case, please report this as an issue:
https://github.com/ydataai/ydata-profiling/issues
(include the error message: 'could not convert string to float: 'std'')
  warnings.warn(


Generate report structure:   0%|          | 0/1 [00:00<?, ?it/s]

Render HTML:   0%|          | 0/1 [00:00<?, ?it/s]

Export report to file:   0%|          | 0/1 [00:00<?, ?it/s]

In [26]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 205 entries, 0 to 204
Data columns (total 26 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   symboling          205 non-null    int64  
 1   normalized-losses  205 non-null    object 
 2   make               205 non-null    object 
 3   fuel-type          205 non-null    object 
 4   aspiration         205 non-null    object 
 5   num-of-doors       205 non-null    object 
 6   body-style         205 non-null    object 
 7   drive-wheels       205 non-null    object 
 8   engine-location    205 non-null    object 
 9   wheel-base         205 non-null    float64
 10  length             205 non-null    float64
 11  width              205 non-null    float64
 12  height             205 non-null    float64
 13  curb-weight        205 non-null    int64  
 14  engine-type        205 non-null    object 
 15  num-of-cylinders   205 non-null    object 
 16  engine-size        205 non

<b>How to work with missing data?</b>

Steps for working with missing data:
<ol>
    <li>dentify missing data</li>
    <li>deal with missing data</li>
    <li>correct data format</li>
</ol>

In [39]:
df = df.replace("?", np.NaN)
df.sample(10)

symboling normalized-losses         make fuel-type aspiration  \
65           0               118        mazda       gas        std   
22           1               118        dodge       gas        std   
139          2                83       subaru       gas        std   
127          3               NaN      porsche       gas        std   
123         -1                74     plymouth       gas        std   
79           1               161   mitsubishi       gas      turbo   
0            3               NaN  alfa-romero       gas        std   
185          2                94   volkswagen       gas        std   
63           0               NaN        mazda    diesel        std   
165          1               168       toyota       gas        std   

    num-of-doors   body-style drive-wheels engine-location  wheel-base  ...  \
65          four        sedan          rwd           front       104.9  ...   
22           two    hatchback          fwd           front        93.7  ...   
139          two    hatchback          fwd           front        93.7  ...   
127          two      hardtop          rwd            rear        89.5  ...   
123         four        wagon          fwd           front       103.3  ...   
79           two    hatchback          fwd           front        93.0  ...   
0            two  convertible          rwd           front        88.6  ...   
185         four        sedan          fwd           front        97.3  ...   
63           NaN        sedan          fwd           front        98.8  ...   
165          two        sedan          rwd           front        94.5  ...   

     engine-size  fuel-system  bore  stroke compression-ratio horsepower  \
65           140         mpfi  3.76    3.16               8.0        120   
22            90         2bbl  2.97    3.23               9.4         68   
139          108         2bbl  3.62    2.64               8.7         73   
127          194         mpfi  3.74    2.90               9.5        207   
123          122         2bbl  3.35    3.46               8.5         88   
79            98         spdi  3.03    3.39               7.6        102   
0            130         mpfi  3.47    2.68               9.0        111   
185          109         mpfi  3.19    3.40               9.0         85   
63           122          idi  3.39    3.39              22.7         64   
165           98         mpfi  3.24    3.08               9.4        112   

     peak-rpm city-mpg highway-mpg  price  
65       5000       19          27  18280  
22       5500       31          38   6377  
139      4400       26          31   7053  
127      5900       17          25  34028  
123      5000       24          30   8921  
79       5500       24          30   7689  
0        5000       21          27  13495  
185      5250       27          34   8195  
63       4650       36          42  10795  
165      6600       26          29   9298  

[10 rows x 26 columns]

In [43]:
df.isna().sum().sum()

59

## Identify missing values

In [46]:
df.isnull().any().any()

True

In [47]:
df.isnull().sum()

symboling             0
normalized-losses    41
make                  0
fuel-type             0
aspiration            0
num-of-doors          2
body-style            0
drive-wheels          0
engine-location       0
wheel-base            0
length                0
width                 0
height                0
curb-weight           0
engine-type           0
num-of-cylinders      0
engine-size           0
fuel-system           0
bore                  4
stroke                4
compression-ratio     0
horsepower            2
peak-rpm              2
city-mpg              0
highway-mpg           0
price                 4
dtype: int64

Based on the summary above, each column has 205 rows of data, seven columns containing missing data:
<ol>
    <li>"normalized-losses": 41 missing data</li>
    <li>"num-of-doors": 2 missing data</li>
    <li>"bore": 4 missing data</li>
    <li>"stroke" : 4 missing data</li>
    <li>"horsepower": 2 missing data</li>
    <li>"peak-rpm": 2 missing data</li>
    <li>"price": 4 missing data</li>
</ol>

<h3 id="deal_missing_values">Deal with missing data</h3>
<b>How to deal with missing data?</b>

<ol>
    <li>drop data<br>
        a. drop the whole row<br>
        b. drop the whole column
    </li>
    <li>replace data<br>
        a. replace it by mean<br>
        b. replace it by frequency<br>
        c. replace it based on other functions
    </li>
</ol>

Whole columns should be dropped only if most entries in the column are empty. In our dataset, none of the columns are empty enough to drop entirely.
We have some freedom in choosing which method to replace data; however, some methods may seem more reasonable than others. We will apply each method to many different columns:

<b>Replace by mean:</b>
<ul>
    <li>"normalized-losses": 41 missing data, replace them with mean</li>
    <li>"stroke": 4 missing data, replace them with mean</li>
    <li>"bore": 4 missing data, replace them with mean</li>
    <li>"horsepower": 2 missing data, replace them with mean</li>
    <li>"peak-rpm": 2 missing data, replace them with mean</li>
</ul>

<b>Replace by frequency:</b>
<ul>
    <li>"num-of-doors": 2 missing data, replace them with "four". 
        <ul>
            <li>Reason: 84% sedans is four doors. Since four doors is most frequent, it is most likely to occur</li>
        </ul>
    </li>
</ul>

<h4>Calculate the average of the column </h4>

In [48]:
avg_norm_loss = df['normalized-losses'].astype("float").mean()
avg_norm_loss

122.0

<h4>Replace "NaN" by mean value in "normalized-losses" column</h4>

In [49]:
df["normalized-losses"] = df["normalized-losses"].replace(np.NaN, avg_norm_loss)
df["normalized-losses"]

0      122.0
1      122.0
2      122.0
3        164
4        164
       ...  
200       95
201       95
202       95
203       95
204       95
Name: normalized-losses, Length: 205, dtype: object

<h4>Calculate the mean value for 'bore' column</h4>

In [50]:
avg_bore = df["bore"].astype("float").mean()
df['bore'] = df["bore"].replace(np.NaN, avg_bore)


In [51]:
avg_stroke = df["stroke"].astype("float").mean(axis=0)
print("Average of stroke:", avg_stroke)

# replace NaN by mean value in "stroke" column
df["stroke"] = df["stroke"].replace(np.nan, avg_stroke)
df["stroke"]

Average of stroke: 3.255422885572139


0      2.68
1      2.68
2      3.47
3      3.40
4      3.40
       ... 
200    3.15
201    3.15
202    2.87
203    3.40
204    3.15
Name: stroke, Length: 205, dtype: object

In [52]:
df.isnull().sum()

symboling            0
normalized-losses    0
make                 0
fuel-type            0
aspiration           0
num-of-doors         2
body-style           0
drive-wheels         0
engine-location      0
wheel-base           0
length               0
width                0
height               0
curb-weight          0
engine-type          0
num-of-cylinders     0
engine-size          0
fuel-system          0
bore                 0
stroke               0
compression-ratio    0
horsepower           2
peak-rpm             2
city-mpg             0
highway-mpg          0
price                4
dtype: int64

In [54]:
avg_horsepower = df['horsepower'].astype('float').mean(axis=0)
print("Average horsepower:", avg_horsepower)
df['horsepower'] = df['horsepower'].replace(np.nan, avg_horsepower)

Average horsepower: 104.25615763546797


In [55]:
avg_peakrpm= df['peak-rpm'].astype('float').mean(axis=0)
print("Average peak rpm:", avg_peakrpm)
df['peak-rpm'] = df['peak-rpm'].replace(np.nan, avg_peakrpm)

Average peak rpm: 5125.369458128079


To see which values are present in a particular column, we can use the ".value_counts()" method:

In [76]:
df['num-of-doors'].value_counts()

num-of-doors
four    114
two      89
Name: count, dtype: int64

We can see that four doors are the most common type. We can also use the ".idxmax()" method to calculate for us the most common type automatically:

In [77]:
max_category = df["num-of-doors"].value_counts().idxmax()
max_category

'four'

In [79]:
# replace the missing 'num-of-doors' values by the most frequent
df["num-of-doors"] = df["num-of-doors"].replace(np.nan, max_category)

<h4>Lets list the data types for each column</h4>

In [81]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 205 entries, 0 to 204
Data columns (total 26 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   symboling          205 non-null    int64  
 1   normalized-losses  205 non-null    object 
 2   make               205 non-null    object 
 3   fuel-type          205 non-null    object 
 4   aspiration         205 non-null    object 
 5   num-of-doors       205 non-null    object 
 6   body-style         205 non-null    object 
 7   drive-wheels       205 non-null    object 
 8   engine-location    205 non-null    object 
 9   wheel-base         205 non-null    float64
 10  length             205 non-null    float64
 11  width              205 non-null    float64
 12  height             205 non-null    float64
 13  curb-weight        205 non-null    int64  
 14  engine-type        205 non-null    object 
 15  num-of-cylinders   205 non-null    object 
 16  engine-size        205 non

<h4>Convert data types to proper format</h4>

In [82]:
df[["bore", "stroke"]] = df[["bore", "stroke"]].astype("float")
df[["normalized-losses"]] = df[["normalized-losses"]].astype("int")
df[["price"]] = df[["price"]].astype("float")
df[["peak-rpm"]] = df[["peak-rpm"]].astype("float")
df.head()

symboling  normalized-losses         make fuel-type aspiration  \
0          3                122  alfa-romero       gas        std   
1          3                122  alfa-romero       gas        std   
2          1                122  alfa-romero       gas        std   
3          2                164         audi       gas        std   
4          2                164         audi       gas        std   

  num-of-doors   body-style drive-wheels engine-location  wheel-base  ...  \
0          two  convertible          rwd           front        88.6  ...   
1          two  convertible          rwd           front        88.6  ...   
2          two    hatchback          rwd           front        94.5  ...   
3         four        sedan          fwd           front        99.8  ...   
4         four        sedan          4wd           front        99.4  ...   

   engine-size  fuel-system  bore  stroke compression-ratio horsepower  \
0          130         mpfi  3.47    2.68               9.0        111   
1          130         mpfi  3.47    2.68               9.0        111   
2          152         mpfi  2.68    3.47               9.0        154   
3          109         mpfi  3.19    3.40              10.0        102   
4          136         mpfi  3.19    3.40               8.0        115   

   peak-rpm city-mpg  highway-mpg    price  
0    5000.0       21           27  13495.0  
1    5000.0       21           27  16500.0  
2    5000.0       19           26  16500.0  
3    5500.0       24           30  13950.0  
4    5500.0       18           22  17450.0  

[5 rows x 26 columns]